File: backend/src/main/java/ch/rfobaden/incidentmanager/backend/models/Incident.java
package ch.rfobaden.incidentmanager.backend.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Table;

/**
 * {@code Incident} represents any major event handled via the IncidentManager.
 * It is further divided into {@link Transport} and {@link Report} entities.
 */
@Entity
@Table(name = "incident")
public class Incident extends ClosableModel
    implements ImageOwner, DocumentOwner {

    /**
     * The reason for closing the incident.
     * It is {@code null} if the incident has never been {@link ClosableModel#isClosed() closed}.
     */
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "close_reason_id")
    private CloseReason closeReason;

    /**
     * The {@link Report reports} of the incident.
     */
    @OneToMany(mappedBy = "incident", cascade = CascadeType.REMOVE)
    private List<Report> reports = new ArrayList<>();


    /**
     * The {@link Transport transports} of the incident.
     */
    @OneToMany(mappedBy = "incident", cascade = CascadeType.REMOVE)
    private List<Transport> transports = new ArrayList<>();

    /**
     * The images attached to the incident, stored as {@link Document} instances.
     */
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Document> images = new ArrayList<>();

    /**
     * The {@link Document documents} attached to the incident.
     * Does not include the incident's {@link #images image documents}.
     */
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Document> documents = new ArrayList<>();

    public CloseReason getCloseReason() {
        return closeReason;
    }

    public void setCloseReason(CloseReason closeReason) {
        this.closeReason = closeReason;
    }

    /**
     * Whether the incident is done.
     * An incident is done when all its {@link #getTransports() transports}
     * and {@link #getReports() reports} are all closed or done.
     *
     * @return Whether the incident is done.
     */
    @JsonProperty("isDone")
    public boolean isDone() {
        return (!getTransports().isEmpty() || !getReports().isEmpty())
            && getTransports().stream().allMatch(Transport::isClosed)
            && (getReports().stream().allMatch(Report::isClosed)
                || getReports().stream().allMatch(Report::isDone));
    }

    @JsonIgnore
    public List<Report> getReports() {
        return reports;
    }

    /**
     * Lists the {@link Report#getId() ids} of all closed {@link #getReports() reports}.
     *
     * @return The ids of all closed reports.
     */
    public List<Long> getClosedReportIds() {
        return getReports().stream()
            .filter(r -> r.isClosed() || r.isDone())
            .map(Report::getId)
            .collect(Collectors.toList());
    }

    /**
     * Lists the {@link Report#getId() ids} of all {@link #getReports()} reports}.
     *
     * @return The ids of all reports.
     */
    public List<Long> getReportIds() {
        return getReports().stream().map(Report::getId).collect(Collectors.toList());
    }

    @JsonIgnore
    public void setReports(List<Report> reports) {
        this.reports = reports;
    }

    @JsonIgnore
    public List<Transport> getTransports() {
        return transports;
    }

    @JsonIgnore
    public void setTransports(List<Transport> transports) {
        this.transports = transports;
    }

    /**
     * Lists the {@link Transport#getId() ids} of all {@link #getTransports() transports}.
     *
     * @return The ids of all transports.
     */
    public List<Long> getTransportIds() {
        return getTransports().stream().map(Transport::getId).collect(Collectors.toList());
    }
    
    /**
     * Lists the {@link Report#getId() ids} of all closed {@link #getTransports() transports}.
     *
     * @return The ids of all closed transports.
     */
    public List<Long> getClosedTransportIds() {
        return getTransports().stream()
            .filter(Transport::isClosed)
            .map(Transport::getId)
            .collect(Collectors.toList());
    }

    @Override
    public List<Document> getImages() {
        return images;
    }

    @Override
    public List<Document> getDocuments() {
        return documents;
    }

    @Override
    public void setDocuments(List<Document> documents) {
        this.documents = documents;
    }

    @Override
    public void setImages(List<Document> images) {
        this.images = images;
    }

    /**
     * Creates the set of the ids of all {@link Organization organizations}
     * that are connected to this incident.
     *
     * @return The set of ids of all connected organizations.
     */
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    public Set<Long> getOrganizationIds() {
        return Stream.concat(
                Stream.concat(
                    Stream.concat(
                        reports.stream(),
                        transports.stream()
                    ),
                    reports.stream()
                        .map(Report::getTasks)
                        .flatMap(Collection::stream)
                ),
                reports.stream()
                    .map(Report::getTasks)
                    .flatMap(Collection::stream)
                    .map(Task::getSubtasks)
                    .flatMap(Collection::stream)
            )
            .map(Trackable::getAssignee)
            .filter(Objects::nonNull)
            .map(User::getOrganizationId)
            .filter(Objects::nonNull)
            .collect(Collectors.toUnmodifiableSet());
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Incident)) {
            return false;
        }
        var incident = (Incident) other;
        return equalsClosableModel(other)
            && Objects.equals(closeReason, incident.closeReason);
    }

    @Override
    public int hashCode() {
        return Objects.hash(closableModelHashCode(), closeReason);
    }
}

File: frontend/models/Incident.ts
import Model, { parseModel } from '@/models/base/Model'
import { parseDate } from '@/models/base/Date'
import CloseReason, { parseCloseReason } from '@/models/CloseReason'
import Report from '@/models/Report'
import Transport from '@/models/Transport'
import Id from '@/models/base/Id'
import Document from '@/models/Document'
import Organization from '@/models/Organization'

/**
 * `Incident` represents any major event handled via the IncidentManager.
 * It is further divided into {@link Transport} and {@link Report} entities.
 */
export default interface Incident extends Model {
  /**
   * The title of the incident.
   */
  title: string

  /**
   * A textual description of what the incident is about.
   */
  description: string | null

  /**
   * The moment in time at which the incident will start.
   * This represents the actual time at which the real-life event
   * managed in this entity will start.
   * <p>
   *     This is used to plan an incident in advance.
   * </p>
   */
  startsAt: Date | null

  /**
   * The moment in time at which the incident will end.
   * This represents the actual time at which the real-life event
   * managed in this entity will end.
   */
  endsAt: Date | null

  /**
   * The reason for closing the incident.
   * It is {@code null} if the incident has never been {@link #isClosed closed}.
   */
  closeReason: CloseReason | null

  /**
   * Whether the incident is closed.
   * A closed incident counts as completed.
   */
  isClosed: boolean

  /**
   * Whether the incident is done.
   * An incident is done when all its {@link Transport transports}
   * and {@link Report reports} are all closed or done.
   */
  isDone: boolean

  /**
   * Lists the ids of all closed {@link Report reports}.
   */
  closedReportIds: Id<Report>[]

  /**
   * Lists the ids of all {@link Report reports}.
   */
  reportIds: Id<Report>[]

  /**
   * Lists the ids of all closed {@link Transport transports}.
   */
  closedTransportIds: Id<Transport>[]

  /**
   * Lists the ids of all {@link Transport transports}.
   */
  transportIds: Id<Transport>[]

  /**
   * Set of ids of all {@link Organization organizations}
   * that are connected to this incident.
   */
  organizationIds: Id<Organization>[]

  /**
   * The images attached to the incident, stored as {@link Document} instances.
   */
  images: Document[]

  /**
   * The {@link Document documents} attached to the incident.
   * Does not include the incident's {@link #images image documents}.
   */
  documents: Document[]
}

/**
 * Parses an incident's JSON structure.
 *
 * @param data The incident to parse.
 * @return The parsed incident.
 */
export const parseIncident = (data: Incident): Incident => {
  return {
    ...data,
    ...parseModel(data),
    startsAt: parseDateOrNull(data.startsAt),
    endsAt: parseDateOrNull(data.endsAt),
    closeReason: data.closeReason && parseCloseReason(data.closeReason),
  }
}

/**
 * Parses a date's JSON structure.
 *
 * @param date The date to parse.
 * @return The parsed date.
 */
const parseDateOrNull = (date: Date | null): Date | null => {
  return date === null ? null : parseDate(date)
}

/**
 * `ClosedIncident` represents an incident that is {@link Incident.isClosed closed}.
 */
export interface ClosedIncident extends Incident {
  closeReason: CloseReason
  isClosed: true
}

/**
 * Checks if an incident is an instance of {@link ClosedIncident}.
 *
 * @param incident The incident.
 * @return Whether the incident is a {@link ClosedIncident}.
 */
export const isClosedIncident = (incident: Incident): incident is ClosedIncident =>
  incident.isClosed && incident.closeReason !== null
